### PR TITLE
Feat (brevitas_examples/llm): support for distillation loss

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR

Extend trainer to support distillation loss

## Results

SmolLM2-135M, MXFP4 for weights/activations

| Loss Type      | LR       | PPL      |     Zero-shot    |
| ------------- | ------------- |------------- | ------------- |
| KL (topk=1000) | 1.0 |  31.4 |  37.17 |
| KL | 3.0 |  29.057 |  37.56 |
| CE |  1.5 (default) |  30.38 |  35.99 |
| CE |  2.0 |  27.74 |  37.5 |

### CODE REVIEW

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 1]
```diff
@@ -28,6 +30,128 @@ class TrainingArguments(transformers.TrainingArguments):
...
+    # The distillation terms is given a weight of 1 - \gamma
+    gamma: float = field(default=0.1)
+    # Softmax temperature for the soft targets
+    temperature: float = field(default=1.0)
+    # Interpolation coefficient between 0 and 1, in the generalized
+    # JS divergence
+    beta: float = field(default=0.5)
+
+
+class GeneralizedTrainer(Trainer):
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py:47]: Add License to https://github.com/huggingface/trl/blob/v0.11.4/trl/trainer/gkd_trainer.py.

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 2]
```diff
@@ -28,6 +30,128 @@ class TrainingArguments(transformers.TrainingArguments):
...
+            # Compute the distillation loss
+            distill_loss = GeneralizedTrainer.generalized_jsd_loss(
+                student_logits=outputs.logits,
+                teacher_logits=fp_ouputs.logits,
+                beta=self.beta,
+                temperature=self.temperature,
+            )
+            loss = self.gamma * loss + (1. - self.gamma) * distill_loss
+
+        if (self.args.average_tokens_across_devices and
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: Note that the original loss computed at line 131 (`loss = super().compute_loss(model, inputs, return_outputs, num_items_in_batch)`) has already been scaled by `self.accelerator.num_processes`. Therefore, this `if` statement needs to be only applied to `distill_loss`, i.e.:
```
distill_loss = GeneralizedTrainer.generalized_jsd_loss(...)
if (self.args.average_tokens_across_devices and ...):
    distill_loss *= self.accelerator.num_processes
loss = self.gamma * loss + (1. - self.gamma) * distill_loss
```

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 3]
```diff
@@ -28,6 +30,143 @@ class TrainingArguments(transformers.TrainingArguments):
...
+        distillation_return_outputs = return_outputs if not self.use_distillation_loss else True
+
+        loss = super().compute_loss(model, inputs, distillation_return_outputs, num_items_in_batch)
+
+        if distillation_return_outputs:
+            loss, outputs = loss
+
+        if self.use_distillation_loss:
+            with torch.no_grad(), quantization_status_manager(model, disable_act_quant=True, disable_weight_quant=True, disable_bias_quant=True):
+                fp_ouputs = model(**inputs)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]:
```suggestion
                fp_outputs = model(**inputs)
```

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 4]
```diff
@@ -28,6 +30,77 @@ class TrainingArguments(transformers.TrainingArguments):
     # from a checkpoint, so related files are not save by default
     save_strategy: Optional[str] = field(default="no")
 
+    ### Distillation Loss args
+    # Whether to compute the distillation loss
+    use_distillation_loss: bool = field(default=False)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: Maybe move the explanatory comment to the `metadata` of the field, e.g.
```
use_distillation_loss: bool = field(default=False, metadata={"help": "Whether to compute the distillation loss."})
```

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 5]
```diff
@@ -28,6 +30,77 @@ class TrainingArguments(transformers.TrainingArguments):
...
+    # Weight given to the CE loss term in the overall loss
+    # The distillation terms is given a weight of 1 - \gamma
+    gamma: float = field(default=0.1)
+    # Softmax temperature for the soft targets
+    temperature: float = field(default=1.0)
+    topk: int = field(default=-1)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: Add comment for this parameter.

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: We are not using metadata for any of the others, and some of them can get very verbose.

I'd rather keep it as comment for now since we aren't taking advantage of the `metadata` field any way

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 6]
```diff
@@ -28,6 +30,77 @@ class TrainingArguments(transformers.TrainingArguments):
...
+
+        # Compute log probabilities for student and probabilities for teacher
+        student_log_probs = F.log_softmax(student_logits, dim=-1)
+        teacher_probs = F.softmax(teacher_logits, dim=-1)  #.gather(-1, indices).flatten(0, -2)
+        student_log_probs = student_log_probs  #.gather(-1, indices).flatten(0, -2)
+
+        if topk > 0:
+            teacher_probs = teacher_probs.gather(-1, indices).flatten(0, -2)
+            student_log_probs = student_log_probs.gather(-1, indices).flatten(0, -2)
+        jsd = F.kl_div(student_log_probs, teacher_probs, reduction=reduction)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: For numerical stability, I would use `teacher_log_probs = F.log_softmax(teacher_logits, dim=-1)` and compute the loss with `log_target=True` (see https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.kl_div.html). Also, consider renaming the variable, as it is not a general Shannon divergence anymore. I would go for `forward_kl` to make it explicit which loss this is.
```
forward_kld = F.kl_div(student_log_probs, teacher_log_probs, reduction=reduction, log_target=True)
```

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 7]
```diff
@@ -28,6 +30,77 @@ class TrainingArguments(transformers.TrainingArguments):
...
+        student_logits = student_logits / temperature
+        teacher_logits = teacher_logits / temperature
+
+        # Compute log probabilities for student and probabilities for teacher
+        student_log_probs = F.log_softmax(student_logits, dim=-1)
+        teacher_probs = F.softmax(teacher_logits, dim=-1)  #.gather(-1, indices).flatten(0, -2)
+        student_log_probs = student_log_probs  #.gather(-1, indices).flatten(0, -2)
+
+        if topk > 0:
+            teacher_probs = teacher_probs.gather(-1, indices).flatten(0, -2)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: There is no need to gather on this tensor, as this has already been calculated in `torch_ori_logits`. Also, I would do `teacher_logits, indices = teacher_logits.topk(topk, dim=-1, sorted=False)` to ensure that all the subsequent computations are done in the smaller tensor.

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 8]
```diff
@@ -28,6 +30,77 @@ class TrainingArguments(transformers.TrainingArguments):
...
+class GeneralizedTrainer(Trainer):
+
+    def __init__(self, args: TrainingArguments = None, **kwargs) -> None:
+        super().__init__(args=args, **kwargs)
+        self.use_distillation_loss = args.use_distillation_loss
+        self.gamma = args.gamma
+        self.temperature = args.temperature
+
+    @staticmethod
+    def generalized_jsd_loss(
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: Maybe rename to `forward_kl_loss`, or `teacher_student_kl_loss`.

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 9]
```diff
@@ -28,6 +30,79 @@ class TrainingArguments(transformers.TrainingArguments):
     # from a checkpoint, so related files are not save by default
     save_strategy: Optional[str] = field(default="no")
 
+    ### Distillation Loss args
+    use_distillation_loss: bool = field(
+        default=False, metadata={"help": "Whether to compute the distillation loss."})
+    gamma: float = field(
+        default=1., metadata={"help": "Gamma balances CE loss (gamma) vs KE loss (1-gamma)."})
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]:
```suggestion
        default=1., metadata={"help": "Gamma balances CE loss (gamma) vs KL loss (1-gamma)."})
```

File: src/brevitas_examples/llm/llm_quant/rotation_optimization.py [hunk 10]
```diff
@@ -28,6 +30,79 @@ class TrainingArguments(transformers.TrainingArguments):
...
+        # Apply temperature scaling
+        student_logits = student_logits / temperature
+        teacher_logits = teacher_logits / temperature
+
+        # Compute log probabilities for student and probabilities for teacher
+        student_log_probs = F.log_softmax(student_logits, dim=-1)
+        teacher_log_probs = F.log_softmax(teacher_logits, dim=-1)
+        student_log_probs = student_log_probs
+
+        jsd = F.kl_div(student_log_probs, teacher_log_probs, reduction=reduction, log_target=True)
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/rotation_optimization.py]: Nit: return `F.kl_div(...)` directly or use a different name for the variable.

**Review by pablomlago — APPROVED** [review]
LGTM (assuming test pass). I might consider adding an end-to-end test for the distillaiton loss in `LLMRotationOptimizationCases`.
